### PR TITLE
Add Discord Rich Presence as an option

### PR DESCRIPTION
**Pull request by ijre**
This PR adds Discord Rich Presence to the MainForm as an opt-in setting inside the Config -> Customize settings, displaying the current game name and console  
  
Shown below is how it looks:
Discord_ 07, 10, le 31 sur Monday

### CODE REVIEW

**Comment by YoshiRulz** [comment]
Clean your diff please, you don't need +2k/-2k LOC of whitespace changes in Designer files. If you recreate the changes on a new branch it'd fix that plus there wouldn't be merge commits. Force push with `git push -f origin new-branch:master` (origin being your fork) to overwrite this PR.

**Comment by ijre** [comment]
Ah, my bad; I'll fix that yeah

**Comment by nattthebear** [comment]
What if anything does this do on linux builds?

**Comment by adelikat** [comment]
if someone can fix the merge conflicts, I will merge this feature

File: src/BizHawk.Client.EmuHawk/MainForm.cs [hunk 1]
```diff
@@ -49,6 +51,8 @@ public partial class MainForm : Form, IMainFormForApi, IMainFormForConfig, IMain
 			(new[] { "PCE", "PCECD", "SGX" }, new[] { CoreNames.HyperNyma, CoreNames.PceHawk, CoreNames.TurboNyma })
 		};
 
+		private static DiscordRpcClient Client = new DiscordRpcClient("749686424248385716");
```
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.cs:54]: Is this a token generated for EmuHawk? Did you need to register as a developer/partner/whatever to get one?

File: src/BizHawk.Client.EmuHawk/MainForm.cs [hunk 2]
```diff
@@ -4461,6 +4476,36 @@ private bool Rewind(ref bool runFrame, long currentTimestamp, out bool returnToR
...
+					Assets = new Assets { LargeImageKey = "corphawk" },
+					Details = Emulator.IsNull() ? "No ROM loaded"
+					:
+					$"{Client.CurrentUser.Username} is playing {Game.Name} for the {Emulator.System().DisplayName}"
+				});
+		}
+
+		private void SetupPresence()
+		{
+			if (Process.GetProcessesByName("Discord").Length == 0) // check to make sure discord is actually open
```
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.cs:4493]: ew

File: src/BizHawk.Client.EmuHawk/MainForm.Designer.cs [hunk 3]
```diff
@@ -17,2407 +17,2407 @@ partial class MainForm
 		/// </summary>
 		private void InitializeComponent()
 		{
-			this.components = new System.ComponentModel.Container();
```
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/MainForm.Designer.cs:20]: I still want don't want this checked in. If `git diff -w filename` is blank, don't stage the file.

File: src/BizHawk.Client.EmuHawk/BizHawk.Client.EmuHawk.csproj [hunk 4]
```diff
@@ -85,6 +85,7 @@
 	</ItemGroup>
 -->
 	<ItemGroup>
+		<PackageReference Include="DiscordRichPresence" Version="1.0.150" />
```
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/BizHawk.Client.EmuHawk.csproj:88]: This is a third-party wrapper over the first-party C/C++ lib. Not necessarily a bad thing, I'm just personally wary of adding more technical debt. Also, it targets .NET Standard 2.0.

---

Not sure where to leave this comment, but it'd be nice if this feature was shared between front-ends. Sadly we don't have a good abstraction for that yet, so if this was all moved to `Client.Common`, there wouldn't be an obvious place for it.

File: src/BizHawk.Client.EmuHawk/config/EmuHawkOptions.cs [hunk 5]
```diff
@@ -69,6 +69,7 @@ private void GuiOptions_Load(object sender, EventArgs e)
 			RunInBackgroundCheckbox.Checked = _config.RunInBackground;
 			AcceptBackgroundInputCheckbox.Checked = _config.AcceptBackgroundInput;
 			AcceptBackgroundInputControllerOnlyCheckBox.Checked = _config.AcceptBackgroundInputControllerOnly;
+			DiscordRPC.Checked = _config.DiscordRPC;
```
Comment (YoshiRulz) [inline src/BizHawk.Client.EmuHawk/config/EmuHawkOptions.cs:72]: I think I mentioned this in either IRC or Discord when you were there discussing this feature: there's no precedent for this in the codebase (see #2235). I don't think we should hide this setting in `Config` > `Customize...` with the miscellany. It needs its own `TabPage` at the very least, and probably deserves its own dialog with some warning text in bold.

**Review by YoshiRulz — CHANGES_REQUESTED** [review]
Yo @ijre if you're still AWOL come Monday I'm adding presence myself and closing this PR

**Comment by YoshiRulz** [comment]
I said I'd do this myself, but when I went to find docs I was greeted with:
> The SDK that this documentation references, Discord-RPC, has been deprecated in favor of our new Discord GameSDK.

So this needs to be implemented from scratch. At least there's a first-party library for Unity we can use. When adding Discord integration, we should consider how downstream projects such as multiplayer randomisers could use ApiHawk for networking and be automatically integrated with Discord's "party" UX. All things considered, I think those features can wait until NextHawk.

File: src/BizHawk.Client.EmuHawk/BizHawk.Client.EmuHawk.csproj (hunk 4 above)
Comment (CasualPokePlayer) [reply src/BizHawk.Client.EmuHawk/BizHawk.Client.EmuHawk.csproj:88]: It isn't, this library re-implements the Discord IPC in pure managed C#. It is completely third party in that regard, and not deprecated unlike the C++ library.

I've personally used it (well, [a fork](https://github.com/CasualPokePlayer/GSE/tree/ee07f2b/externals/DiscordRPC)) to implement Discord Rich Presence in my own emulator frontend. It works well enough, the actual IPC protocol is quite simple, other than Discord having the genius decision of using JSON for IPC messages.

Its JSON usage also comes with a minor wrinkle: this library uses Newtonsoft.Json, which of course has issues for trimming/NativeAOT in the future, along with known footguns (but that probably doesn't matter too much). My fork specifically swaps it out with System.Text.Json, which isn't that hard to do, but would be trickier to add to upstream due to some of their wants (compat with ancient .NET versions / some ancient Unity crap which needs Newtonsoft.Json).

**Comment by CasualPokePlayer** [comment]
> The SDK that this documentation references, Discord-RPC, has been deprecated in favor of our new Discord GameSDK.

Discord-RPC was the old official implementation of the Discord IPC protocol (which mind you, the DiscordRichPresence library did not use). This "Discord GameSDK" is another official implementation, having "more features" which also ended up mostly becoming deprecated, so in practice it's not that different of a same feature set compared to the older implementation.

The old Discord-RPC implementation was made in C++ with a C API, and was MIT licensed with full source code available, so emulators generally used it. The Discord GameSDK actually just has functionality locked behind native libraries (probably C++ again), without any available source code and no FOSS license, just API headers. As such, pretty much all emulators have kept to the old Discord-RPC implementation, as many of them cannot use the newer Discord GameSDK as it's just not GPL compatible. Likewise, I don't think we should be using the Discord GameSDK at all. Using the DiscordRichPresence library seems better considering its wide usage and not deprecated status and not needing to pull in native libraries.

File: src/BizHawk.Client.EmuHawk/MainForm.cs (hunk 1 above)
Comment (CasualPokePlayer) [reply src/BizHawk.Client.EmuHawk/MainForm.cs:54]: The client ID is something associated with an "app" on Discord, which can be a bot, have webhooks, do OAuth2 verification, and is used as a front for Discord IPC usage. Any Discord account can become a "developer" for this "app" (although if we want this we'd probably want to create a "team" for BizHawk developers).

**Pull request closed without merging**